accounting/accounting.go

```diff
@@ -0,0 +1,188 @@
+
+	// split the parsed number on the decimal period
+	parts := strings.Split(parsed, ".")
+
+	// in case of exact number...
+	if len(parts) == 1 {
+		return nil
+	}
+
+	// check the float's precision
+	if prec := len([]rune(parts[1])); prec > 2 {
+		return ErrFloatPrecision{
+			Value:     parsed,
+			Precision: prec,
+		}
+	}
+	return nil
+}
+
+// ToMinorUnit returns the currency data in it's minor unit, int64 format
+func ToMinorUnit(c currency.Currency, value float64) int64 {
+	return int64(math.Round(value * float64(c.Factor())))
+}
+
+// FromMinorUnit returns the currency data as a floating point from it's
+// minor currency unit format.
+func FromMinorUnit(c currency.Currency, value int64) float64 {
+	// fast path for currencies like JPY with a factor of 1.
+	if c.Factor() == 1 {
+		return float64(value)
+	}
+	var (
+		v = itof(value)
+		f = ftof(c.FactorAsFloat64())
+	)
+	f64, _ := newf().Quo(v, f).Float64()
+	return f64
+}
+
+// Exchange - Apply currency exchange rates to an amount.
+//
+// value - should always be given in the minor currency unit.
+// exchange - should always be given from the approved finance list.
+//
+// Rounding to the nearest even is a defined business rule.
+// Tills may round up to the nearest penny, but for reporting, the rule is
+// always to use banker's rounding.
+//
+// If unclear, see: // http://wiki.c2.com/?BankersRounding.
+func Exchange(c currency.Currency, value, exchange float64) float64 {
+	// this can happen for example when dealing
+	// with a GBP->GBP exchange, for example.
+	if exchange == 0 {
+		exchange = 1
+	}
+
+	var (
+		v = ftof(value)
+		f = ftof(c.FactorAsFloat64())
+		e = ftof(exchange)
+	)
+	// Here we divide the value, by it's minor currency
+	// unit factor, then divide it once more by the
+	// exchange rate.
+	// -> v / f / e
+	f64, _ := v.Quo(v, f).Quo(v, e).Float64()
+
+	// basic banker's rounding
+	// http://wiki.c2.com/?BankersRounding
+	return math.RoundToEven(f64*100) / 100
+}
+
+// RatNetAmount applies a VAT rate to a big.Rat value. This method returns a big.Float
+// so it's accuracy can be checked, and it's value applied with .Rat(some.field)
+func RatNetAmount(gross, rate *big.Rat) (*big.Float, error) {
+	// Here we go for octuple precision as we are dealing with rational numbers.
+	bf := func(rat *big.Rat) *big.Float {
+		return big.NewFloat(0).SetRat(rat).SetPrec(OctuplePrecision).SetMode(big.ToNearestEven)
+	}
+
+	v := bf(gross)
+	r := bf(rate)
+
+	// Guard against impossible (negative) tax rates.
+	switch r.Cmp(min) {
+	case -1:
+		return min, ErrSubZeroRate
+	case 0:
+		return v, nil
+	}
+	// Turn the rate into a divisor by making it superior to 1.
+	divisor := newf().Add(base, r)
+
+	// Here we divide the gross by it's vat:
+	// -> val / vat
+	// where vat is a gross superior to 1.
+	return newf().Quo(v, divisor), nil
+}
+
+// NetAmount derives the net amount before tax is applied using the given rate.
+func NetAmount(gross int64, rate float64) (int64, error) {
```

Comment:
I would like this method to end up using `RatNetAmount` to make sure we're using the same code for deriving net net amount.

Reply:
done